Add HIL testing

SergioGasquez opened the pull request:
This PR adds the foundations for HIL, It's still in an early phase and might be draft for a while or get merged and get some follow-up PRs.

See the workflow runs in https://github.com/esp-rs/esp-hal/actions/workflows/hil.yml

For instructions on how the testing works, see the [`hil-test` readme](https://github.com/esp-rs/esp-hal/blob/feat/hil-testing/hil-test/README.md)

## Current board support
- Only  C3, C6 and H2 are supported
- C2 should be supported soon
- Xtensa devices would require adding Xtensa support in [`semihosting`](https://github.com/taiki-e/semihosting) and [`probe-rs`](https://github.com/probe-rs/probe-rs/)
  - See https://github.com/taiki-e/semihosting/pull/9 and https://github.com/probe-rs/probe-rs/issues/2301
  - S3 is now supported! We still need to upstream changes and setup a VM with it, so it will probably be in another PR.

## Current tests
- Only `aes`, `gpio`, `uart` and `spi_full_duplex` are partially being tested. See full list of peripherals pending in: https://github.com/esp-rs/esp-hal/issues/809#issuecomment-1999308797

## Todo
- [x] Clear up fixmes and todos in code
- [x] Decide when the workflow needs to run
  - [x] Run when the PR is added to the `merge_queue`  and abort the merge if the CI fails
  - [x] Add a way to manually trigger the workflow with a given repo/branch.
    - Couldn't test those two triggers. Merge queues might require changing some settings, see https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/managing-a-merge-queue#how-merge-queues-work and https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#merge_group
- [ ] Improve documentation on the VMs and wiring in the devkits
  - Atm, we are documenting it on the readme, should we use any other method?
- [x] Find a way to `cfg` away tests for test for targets.
  - Initially though of [`required-features`](https://doc.rust-lang.org/cargo/reference/features.html#required-features), but not really an option since we cant set up multiple chips.
  - There is already a solution, not ideal but does the purpose. See https://github.com/esp-rs/esp-hal/pull/1297#issuecomment-2004320568
- [x] Shall we keep the `log` and `defmt` features or use only one of them? 
  - Removed `log` feature and enabled `defmt` by default: [57e058b](https://github.com/esp-rs/esp-hal/pull/1297/commits/57e058b5dffc19f6e859355218bbb027e2485cff)

Please, feel free to report any feedback or input that you have!

MabezDev commented on hil-test/src/lib.rs:
```diff
@@ -0,0 +1,8 @@
+use defmt_rtt as _;
+use panic_probe as _;
+
+pub mod esp_hal {
```
What's the purpose of this? Could we not just use `pub use esp_hal` here?

SergioGasquez replied:
Ups its a [leftover](https://github.com/esp-rs/esp-hal/pull/1297/commits/1ec70a8bb0c25d7dda2afa7b41cbde9bf697eed0#diff-8ca170be4c58f4bdbe4e779de8359cbe40188475a67baa692f3f84bcd3d9c46fR7-R20) of when we had several crates for hal, I quickly changed it to test it and forgot to update it. Will update it!

MabezDev commented on hil-test/tests/aes.rs:
```diff
@@ -0,0 +1,74 @@
+    }
+}
+
+#[embedded_test::tests]
```
Maybe we can add a cfg here to cfg the tests away? Not sure if it will work though.

SergioGasquez replied:
Tried that earlier, but it fails.

MabezDev replied:
Damn :sob:

MabezDev commented on hil-test/README.md:
```diff
@@ -0,0 +1,108 @@
+
+Hardware-in-loop testing for `esp-hal`.
+
+For assistance with this package please [open an issue] or [start a discussion], or contact @jessebraham on Matrix.
```
We should remove having a specific contact, in favour of sharing the load between us all.

SergioGasquez replied:
Definitely, also slipped in!

MabezDev commented on hil-test/tests/spi_full_duplex.rs:
```diff
@@ -0,0 +1,105 @@
+    fn test_symestric_transfer(mut ctx: Context) {
+        let write = [0xde, 0xad, 0xbe, 0xef];
+        let mut read: [u8; 4] = [0x00u8; 4];
+        SpiBus::transfer(&mut ctx.spi, &mut read[..], &write[..])
```
We should be able to just use the `transfer` method now the traits aren't included in the prelude (the PR might need a rebase).

SergioGasquez commented:
Just pushed https://github.com/esp-rs/esp-hal/pull/1297/commits/700a86b9c181fdacdbbc8a0fee8178b78335a8fc which shows a way to configure away tests of peripherals for targets that dont support that peripheral. Not an ideal solution imo, but couldn't find any other way (thanks @t-moe for the suggestion)

jessebraham commented on hil-test/.cargo/config.toml:
```diff
@@ -0,0 +1,22 @@
+[env]
+DEFMT_LOG="info"
+RUST_LOG = "warn"
```
What's the reason for setting different log levels for the two environment variables here?

SergioGasquez replied:
`probe-rs `was too verbose, but I'll just remove them

jessebraham commented on hil-test/.cargo/config.toml:
```diff
@@ -0,0 +1,22 @@
+rustflags = [
+    "-C", "link-arg=-Tlinkall.x",
+    "-C", "link-arg=-Tembedded-test.x",
+    "-C", "link-arg=-Tdefmt.x", # Comment this line when using log feature
```
Do we need the `log` feature at all? If the testing framework is all based on `defmt` it makes sense to me to only use that, I can't imagine a situation where `log` will ever be required in this package?

SergioGasquez replied:
Definetly `log` is not required, just wanted to confirm that it's not useful neither, that's why I put it in the Todo list of the PR. I'm happy to remove it!

t-moe replied:
Note that you can use log or defmt with embedded-test. I use log in my tests, because core::fmt::Debug is implemented for more types than defmt::Format. Also core has more assert macro variants (e.g. assert_matches) than defmt... But up to you to decide...

jessebraham reviewed (approved):
I think, especially for our first iteration, this is looking really good! Thank you so much for all your work here, this is a huge win for us.

We should probably expand the section in `README.md` regarding adding tests, but that is absolutely beyond this scope of this PR and can be tackled separately.

So I'll say LGTM, but let's wait for confirmation from @MabezDev and/or @bjoernQ as well :)

SergioGasquez commented:
Sorry for adding late changes.
- I stole the alias idea and implemented it here to have a simple way of running all the tests for a target
- Pinned the moving dependencies (`semihosting`, `embedded-test` and `probe-rs`)
- Updated the README

After the pinning/alias changes, I ran the CI to verify that everything was working, here is the run https://github.com/esp-rs/esp-hal/actions/runs/8372136967 (I amended the commit and removed the `push` trigger of the workflow afterward)

bjoernQ reviewed (approved):
LGTM ❤️

MabezDev reviewed (approved):
Thank you @SergioGasquez!

SergioGasquez commented:
Just tested all the targets locally:
- ESP32C3:
<details>
```
❯ cargo esp32c3
warning: method `into_result` is never used
   --> /home/sergio/Documents/Espressif/esp-rs/esp-hal/esp-hal/src/fmt.rs:204:8
    |
201 | pub trait Try {
    |           --- method in this trait
...
204 |     fn into_result(self) -> Result<Self::Ok, Self::Error>;
    |        ^^^^^^^^^^^
    |
    = note: `#[warn(dead_code)]` on by default

warning: `esp-hal` (lib) generated 1 warning
    Finished `release` profile [optimized + debuginfo] target(s) in 0.06s
     Running tests/aes.rs (target/riscv32imc-unknown-none-elf/release/deps/aes-40fe1a1875ed0f4c)
      Erasing ✔ [00:00:00] [#####################################################################################################################################################################################] 192.00 KiB/192.00 KiB @ 239.08 KiB/s (eta 0s )
  Programming ✔ [00:00:00] [########################################################################################################################################################################################] 35.52 KiB/35.52 KiB @ 64.83 KiB/s (eta 0s )    Finished in 1.363s

running 2 tests
test tests::test_aes_encryption ... ok
test tests::test_aes_decryption ... ok

test result: ok. 2 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.87s

     Running tests/gpio.rs (target/riscv32imc-unknown-none-elf/release/deps/gpio-3ace83cd88ceff7c)
      Erasing ✔ [00:00:00] [#####################################################################################################################################################################################] 192.00 KiB/192.00 KiB @ 238.35 KiB/s (eta 0s )
  Programming ✔ [00:00:00] [########################################################################################################################################################################################] 35.26 KiB/35.26 KiB @ 64.59 KiB/s (eta 0s )    Finished in 1.363s

running 2 tests
test tests::test_gpio_input  ... ok
test tests::test_gpio_output ... ok

test result: ok. 2 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.86s

     Running tests/spi_full_duplex.rs (target/riscv32imc-unknown-none-elf/release/deps/spi_full_duplex-28d01ab612584813)
      Erasing ✔ [00:00:00] [#####################################################################################################################################################################################] 192.00 KiB/192.00 KiB @ 240.25 KiB/s (eta 0s )
  Programming ✔ [00:00:00] [########################################################################################################################################################################################] 37.06 KiB/37.06 KiB @ 66.70 KiB/s (eta 0s )    Finished in 1.367s

running 4 tests
test tests::test_symestric_transfer                      ... ok
test tests::test_asymestric_transfer                     ... ok
test tests::test_symestric_transfer_huge_buffer          ... ok
test tests::test_symestric_transfer_huge_buffer_no_alloc ... ok

test result: ok. 4 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 1.72s

     Running tests/uart.rs (target/riscv32imc-unknown-none-elf/release/deps/uart-12ebab303e43539b)
      Erasing ✔ [00:00:00] [#####################################################################################################################################################################################] 192.00 KiB/192.00 KiB @ 236.60 KiB/s (eta 0s )
  Programming ✔ [00:00:00] [########################################################################################################################################################################################] 35.19 KiB/35.19 KiB @ 64.28 KiB/s (eta 0s )    Finished in 1.37s

running 1 test
test tests::test_send_receive ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.44s
```
</details>
- ESP32-C6:
<details>
```
   Compiling hil-test v0.0.0 (/home/sergio/Documents/Espressif/esp-rs/esp-hal/hil-test)
    Finished `release` profile [optimized + debuginfo] target(s) in 6.21s
     Running tests/aes.rs (target/riscv32imac-unknown-none-elf/release/deps/aes-79d2fe8d6b7de728)
      Erasing ✔ [00:00:00] [#####################################################################################################################################################################################] 128.00 KiB/128.00 KiB @ 634.74 KiB/s (eta 0s )
  Programming ✔ [00:00:00] [########################################################################################################################################################################################] 36.33 KiB/36.33 KiB @ 96.69 KiB/s (eta 0s )    Finished in 0.589s

running 2 tests
test tests::test_aes_encryption ... ok
test tests::test_aes_decryption ... ok

test result: ok. 2 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.86s

     Running tests/gpio.rs (target/riscv32imac-unknown-none-elf/release/deps/gpio-92410ef5e9f8728d)
      Erasing ✔ [00:00:00] [#####################################################################################################################################################################################] 128.00 KiB/128.00 KiB @ 646.24 KiB/s (eta 0s )
  Programming ✔ [00:00:00] [########################################################################################################################################################################################] 36.06 KiB/36.06 KiB @ 96.54 KiB/s (eta 0s )    Finished in 0.583s

running 2 tests
test tests::test_gpio_input  ... ok
test tests::test_gpio_output ... ok

test result: ok. 2 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.86s

     Running tests/spi_full_duplex.rs (target/riscv32imac-unknown-none-elf/release/deps/spi_full_duplex-5f2314b91cd414c3)
      Erasing ✔ [00:00:00] [#####################################################################################################################################################################################] 128.00 KiB/128.00 KiB @ 649.48 KiB/s (eta 0s )
  Programming ✔ [00:00:00] [########################################################################################################################################################################################] 37.98 KiB/37.98 KiB @ 97.06 KiB/s (eta 0s )    Finished in 0.6s

running 4 tests
test tests::test_symestric_transfer                      ... ok
test tests::test_asymestric_transfer                     ... ok
test tests::test_symestric_transfer_huge_buffer          ... ok
test tests::test_symestric_transfer_huge_buffer_no_alloc ... ok

test result: ok. 4 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 1.71s

     Running tests/uart.rs (target/riscv32imac-unknown-none-elf/release/deps/uart-456906ed79152f2a)
      Erasing ✔ [00:00:00] [#####################################################################################################################################################################################] 128.00 KiB/128.00 KiB @ 653.25 KiB/s (eta 0s )
  Programming ✔ [00:00:00] [########################################################################################################################################################################################] 36.02 KiB/36.02 KiB @ 96.73 KiB/s (eta 0s )    Finished in 0.579s

running 1 test
test tests::test_send_receive ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.43s
```
</details>
- ESP32-H2:
<details>
```
   Compiling hil-test v0.0.0 (/home/sergio/Documents/Espressif/esp-rs/esp-hal/hil-test)
    Finished `release` profile [optimized + debuginfo] target(s) in 9.01s
     Running tests/aes.rs (target/riscv32imac-unknown-none-elf/release/deps/aes-de854d1506d0a923)
      Erasing ✔ [00:00:00] [#######################################################################################################################################################################################] 192.00 KiB/192.00 KiB @ 1.03 MiB/s (eta 0s )
  Programming ✔ [00:00:00] [########################################################################################################################################################################################] 35.34 KiB/35.34 KiB @ 67.38 KiB/s (eta 0s )    Finished in 0.718s

running 2 tests
test tests::test_aes_encryption ... ok
test tests::test_aes_decryption ... ok

test result: ok. 2 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.86s

     Running tests/gpio.rs (target/riscv32imac-unknown-none-elf/release/deps/gpio-46b476d876631f56)
      Erasing ✔ [00:00:00] [####################################################################################################################################################################################] 192.00 KiB/192.00 KiB @ 1018.32 KiB/s (eta 0s )
  Programming ✔ [00:00:00] [########################################################################################################################################################################################] 35.06 KiB/35.06 KiB @ 66.54 KiB/s (eta 0s )    Finished in 0.727s

running 2 tests
test tests::test_gpio_input  ... ok
test tests::test_gpio_output ... ok

test result: ok. 2 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.86s

     Running tests/spi_full_duplex.rs (target/riscv32imac-unknown-none-elf/release/deps/spi_full_duplex-2b0143ea09f9469e)
      Erasing ✔ [00:00:00] [#######################################################################################################################################################################################] 192.00 KiB/192.00 KiB @ 1.00 MiB/s (eta 0s )
  Programming ✔ [00:00:00] [########################################################################################################################################################################################] 37.05 KiB/37.05 KiB @ 66.30 KiB/s (eta 0s )    Finished in 0.758s

running 4 tests
test tests::test_symestric_transfer                      ... ok
test tests::test_asymestric_transfer                     ... ok
test tests::test_symestric_transfer_huge_buffer          ... ok
test tests::test_symestric_transfer_huge_buffer_no_alloc ... ok

test result: ok. 4 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 1.71s

     Running tests/uart.rs (target/riscv32imac-unknown-none-elf/release/deps/uart-1ce1290e074c134a)
      Erasing ✔ [00:00:00] [#######################################################################################################################################################################################] 192.00 KiB/192.00 KiB @ 1.01 MiB/s (eta 0s )
  Programming ✔ [00:00:00] [########################################################################################################################################################################################] 35.01 KiB/35.01 KiB @ 66.46 KiB/s (eta 0s )    Finished in 0.723s

running 1 test
test tests::test_send_receive ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.43s
```
</details>
- ESP32-S3:
<details>
```
   Compiling hil-test v0.0.0 (/home/sergio/Documents/Espressif/esp-rs/esp-hal/hil-test)
    Finished release [optimized + debuginfo] target(s) in 18.56s
     Running tests/aes.rs (target/xtensa-esp32s3-none-elf/release/deps/aes-e68628e057fe0ffb)
      Erasing ✔ [00:00:00] [#####################################################################################################################################################################################] 192.00 KiB/192.00 KiB @ 336.62 KiB/s (eta 0s )
  Programming ✔ [00:00:00] [########################################################################################################################################################################################] 39.93 KiB/39.93 KiB @ 58.32 KiB/s (eta 0s )    Finished in 1.268s

running 2 tests
test tests::test_aes_encryption ... ok
test tests::test_aes_decryption ... ok

test result: ok. 2 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.84s

     Running tests/gpio.rs (target/xtensa-esp32s3-none-elf/release/deps/gpio-e3b788197b823785)
      Erasing ✔ [00:00:00] [#####################################################################################################################################################################################] 192.00 KiB/192.00 KiB @ 333.05 KiB/s (eta 0s )
  Programming ✔ [00:00:00] [########################################################################################################################################################################################] 39.64 KiB/39.64 KiB @ 58.17 KiB/s (eta 0s )    Finished in 1.27s

running 2 tests
test tests::test_gpio_input  ... ok
test tests::test_gpio_output ... ok

test result: ok. 2 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.85s

     Running tests/spi_full_duplex.rs (target/xtensa-esp32s3-none-elf/release/deps/spi_full_duplex-737c4179a9571d7f)
      Erasing ✔ [00:00:00] [#####################################################################################################################################################################################] 192.00 KiB/192.00 KiB @ 333.78 KiB/s (eta 0s )
  Programming ✔ [00:00:00] [########################################################################################################################################################################################] 41.73 KiB/41.73 KiB @ 59.56 KiB/s (eta 0s )    Finished in 1.289s

running 4 tests
test tests::test_symestric_transfer                      ... ok
test tests::test_asymestric_transfer                     ... ok
test tests::test_symestric_transfer_huge_buffer          ... ok
test tests::test_symestric_transfer_huge_buffer_no_alloc ... ok

test result: ok. 4 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 1.69s

     Running tests/uart.rs (target/xtensa-esp32s3-none-elf/release/deps/uart-0c3c6e9f8846253d)
      Erasing ✔ [00:00:00] [#####################################################################################################################################################################################] 192.00 KiB/192.00 KiB @ 331.86 KiB/s (eta 0s )
  Programming ✔ [00:00:00] [########################################################################################################################################################################################] 39.71 KiB/39.71 KiB @ 58.35 KiB/s (eta 0s )    Finished in 1.272s

running 1 test
test tests::test_send_receive ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.42s
```
